backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java

```diff
@@ -294,21 +293,111 @@ public void setUserIsActive(String username, Boolean active) {
+                .setDescription(roleGroupDescription)
+                .buildAndCreate(_client);
+        _app.createApplicationGroupAssignment(g.getId());
+
+        if (migration) {
+          LOG.info("Created Okta group={} via migration", roleGroupName);
+        }
+
+        if (migration && migrationDestRoles.contains(role)) {
+          migrationDestGroupIds.add(g.getId());
+          migrationDestGroupNames.add(roleGroupName);
+        }
+      } catch (ResourceException e) {
+        if (migration) {
+          // ignore attempts to create groups that already exist if this is in migration mode
+          migrationSourceGroupNames.add(roleGroupName);
+        } else {
+          throw e;
+        }
+      }
+    }
+
+    for (Facility facility : facilities) {
+      String facilityGroupName = generateFacilityGroupName(externalId, facility.getInternalId());
+      String facilityGroupDescription =
+          generateFacilityGroupDescription(name, facility.getFacilityName());
+      try {
+        Group g =
+            GroupBuilder.instance()
+                .setName(facilityGroupName)
+                .setDescription(facilityGroupDescription)
+                .buildAndCreate(_client);
+        _app.createApplicationGroupAssignment(g.getId());
+
+        if (migration) {
+          LOG.info("Created Okta group={} via migration", facilityGroupName);
+        }
+
+      } catch (ResourceException e) {
+        if (migration) {
+          // ignore attempts to create groups that already exist if this is in migration mode
+          migrationSourceGroupNames.add(facilityGroupName);
+        } else {
+          throw e;
+        }
+      }
+    }
+
+    if (migration && !migrationSourceGroupNames.isEmpty() && !migrationDestGroupIds.isEmpty()) {
+      String ruleName = generateMigrationGroupRuleName(externalId, migrationDestRoles);
+      String ruleExpression = generateMigrationGroupRuleExpression(migrationSourceGroupNames);
+      try {
+        // https://developer.okta.com/docs/reference/api/groups/#group-rule-operations
+        GroupRule rule =
+            GroupRuleBuilder.instance()
+                .setType("group_rule")
+                .setName(ruleName)
+                .setGroupRuleExpressionType("urn:okta:expression:1.0")
+                .setGroupRuleExpressionValue(ruleExpression)
+                .setAssignUserToGroups(migrationDestGroupIds)
+                .buildAndCreate(_client);
+        rule.activate();
+
+        LOG.info(
+            "Migrated users from Okta source_groups={} to destination_groups={}",
+            migrationSourceGroupNames,
+            migrationDestGroupNames);
+      } catch (ResourceException e) {
+        LOG.error(
+            "Error migrating users from Okta source_groups={} to destination_groups={}: "
+                + "check if conflicting group_rule={} already exists. "
+                + "Did you delete groups without deleting their relevant group rules?",
```

Comment:
Should we schedule a clean-up to axe those rules at some point?

Reply:
here's the amazing wonderful kicker -- if you delete them programmatically, it undoes all the user migration that those rules performed. if you delete them manually on the Okta UI instead, the migrated users stay where they are. so i'm planning to delete them manually following each stage of the rollout. (if you're not crazed about this idea, i'm going to shrug and point you to Okta.)

Reply:
That's amazing.